specs/anchored-region/anchored-region.md

```diff
@@ -0,0 +1,260 @@
+# Anchored region
+
+## Overview
+
+An *anchored region* is a container component which enables authors to create layouts where the contents of the anchored region can be positioned relative to another "anchor" element.  Additionally, the *anchored region* can react to the available space between the anchor and a parent ["viewport"](https://developer.mozilla.org/en-US/docs/Glossary/viewport) element such that the region is placed on the side of the anchor with the most available space, or even resize itself based on that space.
+
+### Background
+
+This component is inspired by the ["Viewport positioner"](https://github.com/microsoft/fast-dna/issues/2597)  component in the React component set.  It is used as a building block in other components to enable responsive flyouts, or positionable/scaling menus in the [select](https://explore.fast.design/components/select) component. 
+
+A primary goal of the component was to enable authors to create responsive layouts without resorting to expensive [getBoundingClientRect()](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) calls. It instead depends on the more performant [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver) and [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) interfaces.
+
+For a more in-depth understanding of how this component works under the covers please refer to the [intersection observer api](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). 
+
+### Use Cases
+
+It is envisioned that this component would be used as a building block for other components in this library (select, flyout, tooltip, etc) as well as being available for standalone use in responsive layouts.
+
+### Features
+
+- **Relative positioning:** Users can use it to position an element relative to another another element directly, like enabling a menu to open above or below a trigger button.
+
+- **Responsive positioning:** Users can use it to position an element relative to another element based on available space, for example a menu could open upwards if the trigger button is near the bottom of the page, and downwards if it is nearer the top.
+
+- **Responsive scaling:** Users can use it to create a layout region that dynamically sizes depending on space between the anchor and the viewport elements.
```

Comment:
I fear we are putting too many features into this component from the get-go. Can you identify the most basic features needed for this and place any potential future asks in under "Next steps"? See the spec template for details.

Reply:
I think this is the basic set of features based on what functionality that got used in the previous React version.  I have removed the "always on screen" behavior that we had there as it didn't seem to get used.  My reference for what this thing should do is the "position  menu" scenario (select) which the list you flag here is table stakes - menu always opens in same position, menu opens where there is most room, and menu scales with viewport.  There might be some small things we can do without, like maybe the option to only run layout once for example, but I see the proposed feature set as a logical starting point.  We should even be able to steal existing code for a lot of it.

Reply:
So, good point on select, I think it's a good parallel. I'm thinking that select which accepts no options is a good example of default behavior where it merely tries to fit where it can in the available space. Should the anchored region be able to do the same? As in, be able to take no attributes and still have good default behavior, because that is what I think we should be inclined to start with. I agree with including default behavior for what you've listed, "menu opens in same position as default" (like browser select), "menu opens where there is the most room" (like browser select) and I would actually leave the scaling off for now, we could assume it's as big as the user determines it to be since they have full control over the content and styling. In this way we can start off with a smaller component that works out of the box and includes some default behavior and then re-examine every additional feature and add them incrementally.

Reply:
I think we know we want scaling, and we already have working code that I'm feeling reasonably good about being able to transfer (know more soon hopefully) and I want to be sure we support it well so I'd like to see it the initial feature set to be sure it integrates well (I'd want to have scaling prototyped before publishing the initial version at least).  If anything I was thinking that maybe the region *always* scales and authors use css to position items within the scaling region, but I've tended to back away from that as it feels like it would make base implementations more complex because you'd need to worry about css.

Reply:
I agree with the idea of this having a good low-level default that can support the most common scenario. It would be helpful to understand how each of these break down in terms of both high and low-level API. If I have a simple app, I don't want to have to define any properties for this as part of adding a select on my page. For flyout, I may want more configuration, but I think we would similarly support a relatively simple API to start ( top / left) while allowing for more complexity. 

Is there a way that this could be a JS-only utility (or utilities) to support high and low-level API's?

Reply:
I see this one component as a building block for other components and those components would have a lot of flexibility as to how much.  So yes, a select could opt to offer a very basic set of options to the author and the select component would configure its internal anchored region appropriately.  Conversly a flyout could offer more flexibility.  I'm sure this could be reimagined a number of ways, but I think we'd just be moving the complexity around.

Reply:
I think complexity being opted into is desirable, there could be an opportunity for abstraction based on the dependency injections, which I believe is available in some form in the fast-element.

Reply:
Not sure exactly what you mean.  I don't believe the api proposed here adds complexity compared to the fast version.

Reply:
The API is only a small part of my concern, it's the complexity internal to the component (the amount of code we ship with). What I'd like to see is this component being as light as possible out of the box. If there is a way to add responsive scaling, as an example, via dependency injection, that would mean that anyone who is not opting in to responsive scaling doesn't have to carry around the extra bloat. Instead we can create a utility they can inject into the component to provide that functionality if they need it.

Reply:
My primary concern with continuing to churn on this is that it is blocking Select and Auto Suggest components.

If we can't come to a resolution on this quickly I'd like to propose we table this control and prioritize a private implementation for those components so we can get those out the door.